[UI] Add DateTimePicker in UI; fix  clock styles

Manish-Giri opened the pull request:
- Replaced `KeyboardDatePicker` with `KeyboardDateTimePicker` to additionally allow choosing time in pickers across the UI
- Removed `format` prop from three components that render `DatePicker` - `SecretForm`, `ClientForm` and `AuthConsent`, so the new `format` in `DatePicker` works correctly
- Introduced new styles in `theme.js` so that the hour and clock hands display prominently in the pickers, across both dark and light themes

For choosing prominent colors for the clock hands, I started by looking at how the colors were displayed on the [material-ui-pickers demo page](https://material-ui-pickers.dev/demo/datetime-picker). It looks like their clock hand colors match the color of the main toolbar, so I decided to go the same route for the picker here, and chose the color of the main toolbar at `/secrets/create` - `#4177a5`, which is defined as `THEME.SECONDARY` - https://github.com/taskcluster/taskcluster/blob/5e6c2b635b916f495873e9b64fd07a69fb090946/ui/src/utils/constants.js#L52

I used the same color for both dark and light themes, and they stood out quite well, but if required, I can change the color based on theme.

---
## Screenshots (clock visibility) 

### Dark Theme 

### Before - 

![2](https://user-images.githubusercontent.com/11348778/69837278-ebefce00-121b-11ea-8c59-9fb1eab0962e.png)

![3](https://user-images.githubusercontent.com/11348778/69837290-fdd17100-121b-11ea-966d-ab19549d3fea.png)

### After - 

![dark-hour](https://user-images.githubusercontent.com/11348778/69837310-10e44100-121c-11ea-89ea-cca529d9d0ac.png)

![dark-min](https://user-images.githubusercontent.com/11348778/69837348-412bdf80-121c-11ea-9832-a3da20253269.png)

---

### Light Theme

###  Before - 
![5-1](https://user-images.githubusercontent.com/11348778/69837606-ac29e600-121d-11ea-9b52-7aff773f17f6.png)

![6-2](https://user-images.githubusercontent.com/11348778/69837616-b2b85d80-121d-11ea-8d0f-07ecb97c380e.png)

### After -

![light-hour-noop](https://user-images.githubusercontent.com/11348778/69837693-03c85180-121e-11ea-9dc7-747ce80df9a6.png)

![light-hour-noppoint](https://user-images.githubusercontent.com/11348778/69837698-075bd880-121e-11ea-8fa7-84e9435e302e.png)

---

I also changed the background-color of `MuiPickersClockPointer.noPoint`, because otherwise, the white circle in the blue looked a bit jarring -

![light-hour](https://user-images.githubusercontent.com/11348778/69837745-468a2980-121e-11ea-8197-6e619361592f.png)
 

---

Github Bug/Issue: Fixes #2031

Manish-Giri commented:
Looks like I need to add a changelog? Not sure why the other 6 tasks failed.

Manish-Giri commented:
@helfi92 I have fixed your suggestions 2, 3 and 4, but I wasn't able to find this dialog section that has the calendar and time icon in it -

![69886902-56534d80-12b2-11ea-9726-76aadc029cb4](https://user-images.githubusercontent.com/11348778/70382987-c78aa480-1933-11ea-9799-051e92528b1d.png)

I have tried fiddling with various combinations of the picker at `/secrets/create` as well as the demo page at https://material-ui-pickers.dev/demo/datetime-picker, but the dialog box consists of only the date and time in textual form in all cases -

![ww](https://user-images.githubusercontent.com/11348778/70383018-8b0b7880-1934-11ea-8c4f-703e5199e983.png)

I looked at the [KeyboardDateTimePicker API](https://material-ui-pickers.dev/api/KeyboardDateTimePicker) and the only thing that looked like it could enable the dialogbox is the `dateRangeIcon` prop (of type `ReactNode`), but it is currently not used anywhere in `DatePicker.jsx`. Unless I missed a keyboard shortcut that enables the dialog.

If you could tell me the correct way to enable the dialog box, I can then implement your suggestion 1. Thank You.

helfi92 commented:
I'm not sure I understand the question correctly but to change the background color of the dialog that has the calendar and clock icon, we could do something like:

```diff
diff --git a/ui/src/theme.js b/ui/src/theme.js
index cc7559923..98040be50 100644
--- a/ui/src/theme.js
+++ b/ui/src/theme.js
@@ -291,6 +291,11 @@ const createTheme = isDarkTheme => {
           whiteSpace: 'nowrap',
         },
       },
+      MuiPickerDTTabs: {
+        tabs: {
+          backgroundColor: 'red',
+        },
+      },
       MuiPickersToolbar: {
         toolbar: {
           backgroundColor: THEME.SECONDARY,
```

![Screen Shot 2019-12-09 at 3 29 20 PM](https://user-images.githubusercontent.com/3766511/70470298-b1e2bf80-1a98-11ea-9320-4c081658ce08.png)

---

For the background color as well as the color of the tab indicator, we should have something that looks like:

![Screen Shot 2019-12-09 at 3 19 54 PM](https://user-images.githubusercontent.com/3766511/70470244-9a0b3b80-1a98-11ea-8b08-fa8a801fa36c.png)
![Screen Shot 2019-12-09 at 3 20 47 PM](https://user-images.githubusercontent.com/3766511/70470246-9a0b3b80-1a98-11ea-87fe-d1ec9c858203.png)

Hope that helps.

Manish-Giri commented:
@helfi92 Sorry if I wasn't clear enough. What I meant with this - 

> If you could tell me the correct way to enable the dialog box, I can then implement your suggestion 1

was - when I was working on my windows machine on this, I didn't find the section of the dialog box that contains the date and time icons. 

If you notice the screenshots below -

1. *In this screenshot of `/secrets/create`, the dialog box above the actual clock, consists of only the date and time in text*. The section of the dialog that is supposed to consist of the date and time icon (to which you added `background-color: red` in your comment just now) is missing. 
 
![Screenshot 2019-12-07 18 00 11 2](https://user-images.githubusercontent.com/11348778/70471557-6978d100-1a9b-11ea-8786-08927244607d.png)

2. Same thing with the original [datetime-picker component demo](https://material-ui-pickers.dev/demo/datetime-picker) -

![Screenshot 2019-12-07 20 56 04 2](https://user-images.githubusercontent.com/11348778/70471917-2703c400-1a9c-11ea-9aad-d6aea2490c97.png)

But when I open the same demo on my mac machine, the dialog with the date and time icon is present -

![Screenshot_2019-12-09_15_54_58](https://user-images.githubusercontent.com/11348778/70472238-c7f27f00-1a9c-11ea-83e4-0ee2bb8b67ad.png)

I'm not sure what happened there, but I'll go back and investigate tonight. Thank You for your help.

helfi92 commented:
wow. That must be an issue with `@material-ui/pickers`. It would be good if you could file an issue in https://github.com/mui-org/material-ui-pickers/issues.

Manish-Giri commented:
@helfi92 I updated the PR. The latest commit should take care of all your suggestions. 

---

However, that 1 test in `taskcluster-ui` is still failing. Per the task logs, the error message is -
```js
 isDateWithin
    ✔ should return true
    ✖ should return false

...
FAILED TESTS:
  isDateWithin
    ✖ should return false
      Firefox 68.0.0 (Linux 0.0.0)
    expected true to be false
    AssertionError@test/isDateWithin_test.js:169:13
    ./node_modules/chai/lib/chai/assertion.js/module.exports/Assertion.prototype.assert@test/isDateWithin_test.js:478:13
    ./node_modules/chai/lib/chai/core/assertions.js/module.exports/<@test/isDateWithin_test.js:1313:10
    propertyGetter@test/isDateWithin_test.js:8219:29
    proxyGetter@test/isDateWithin_test.js:9729:22
    ./test/isDateWithin_test.js/</</<@test/isDateWithin_test.js:29187:58
    ./test/isDateWithin_test.js/</<@test/isDateWithin_test.js:29186:85
```

On doing some digging, the second test in `isDateWithin_test.js` is failing -
https://github.com/taskcluster/taskcluster/blob/20d15d3ead48f411753037ce8758cf05e7911021/ui/test/isDateWithin_test.js#L19-L29

The `isDateWithin()` function uses `fromNow()` in [fromNow.js](https://github.com/taskcluster/taskcluster/blob/master/ui/src/utils/fromNow.js), which in turn uses [parseTime.js](https://github.com/taskcluster/taskcluster/blob/master/ui/src/utils/parseTime.js). 

Could the error be due to the fact that we replaced DatePicker with DateTimePicker which now has an additional time component? But I'm not sure where these components are being used for the test, (behind the scenes), if that's the case. I didn't find anything in the files I checked.

helfi92 commented:
> Could the error be due to the fact that we replaced DatePicker with DateTimePicker which now has an additional time component?

Nope. That's a known issue https://bugzilla.mozilla.org/show_bug.cgi?id=1598643

Manish-Giri commented:
All right, I'll start working on these two issues. I'll switch to my mac machine because the dialogbox with the calendar and time icons aren't visible on my windows machine. If I could have tested them, I would have perhaps caught the two issues.

> Aside, did you have time to create an issue upstream to raise the issue you were having on windows?

Yes, I created issue https://github.com/mui-org/material-ui-pickers/issues/1419 last night https://github.com/taskcluster/taskcluster/pull/2106#ref-issue-535508079, but they already closed that issue claiming that if the device window height is small, the tabs won't appear, otherwise the picker would need a scrollbar.

My windows machine is a 15.6" laptop, and the browser is already maximized, not sure how I can increase the height further. I also fiddled in dev tools (as they recommended), but no difference.

helfi92 commented:
Interesting. Thanks for filing the issue.

helfi92 commented:
You'll probably want to rebase your work on master so that it doesn't include commits outside this PR.

Manish-Giri commented:
@helfi92 I fixed the last two remaining issues. 

I'm not sure what's going with the additional commits that just got pulled in. I had to change machines to get the calendar and time icons to show up, but I think I forgot to do a `git pull --rebase`, which is why this happened. I'll try it now.

helfi92 commented on ui/src/theme.js:
```diff
@@ -298,7 +298,7 @@ const createTheme = isDarkTheme => {
       MuiPickersToolbarText: {
         toolbarTxt: {
-          color: THEME.PRIMARY_TEXT_DARK,
+          color: 'rgba(255, 255, 255, 0.54)',
```
This would affect everywhere we use a [material-ui tab](https://material-ui.com/components/tabs/#tabs). We'll want to make this change for the pickers component only.

Manish-Giri replied:
Sorry, possible noob question - I did some digging, but I couldn't find the class `.MuiPickersToolbarText-toolbarTxt` being used by `Tab`/`Tabs` (from [material-ui tab](https://material-ui.com/components/tabs/#tabs)) anywhere.  I looked at the classes used in these tabs, and created this [gist](https://gist.github.com/Manish-Giri/19e0f03351521753b155e9e049edff7a) which has all the classes used by `.MuiTab` and `.MuiTabs`.

Were you referring to the color change in the `PrivateTabIndicator` class here?
```css
  PrivateTabIndicator: {
        colorPrimary: {
          backgroundColor: '#fff',
        },
        colorSecondary: {
          backgroundColor: '#000',
        },
      },
```

Because, I found out that while this changes the color of the tab-indicator underneath the calendar-time dialog, as you had shown in your [comment](https://github.com/taskcluster/taskcluster/pull/2106#issuecomment-563424157), this also changes the color of tab-indicator in other places, like in `http://localhost:5080/auth/scopes`, as shown below -

![Screenshot 2019-12-20 21 38 55](https://user-images.githubusercontent.com/11348778/71301902-0a586d80-2373-11ea-96ea-6f285af948e7.png)

Is this what you meant when you said we should be making the change for the Picker component only?

helfi92 replied:
Oh, that's very possible. My mistake.

Manish-Giri replied:
@helfi92 um, so should I try to explore if the color of the TabIndicator can be restricted to the picker component only? For reference, here's how the tabindicator currently looks on https://firefox-ci-tc.services.mozilla.com/auth/scopes -

![Screen Shot 2019-12-23 at 8 22 42 PM](https://user-images.githubusercontent.com/11348778/71387545-6238e380-25c2-11ea-9f17-d6ba5d37f939.png)

But there might be other places too using this tabindicator, which would be affected by this change.

helfi92 replied:
Yes, please :)

Manish-Giri replied:
@helfi92 So, I came across an observation while fiddling with the `PrivateTabIndicator` class - the `span` that has this class also has the `MuiTabs-indicator` class -

![Screenshot 2019-12-24 20 30 53](https://user-images.githubusercontent.com/11348778/71428720-91b62180-2690-11ea-989c-cc7c908f5b3a.png)

If I target this class instead of the `PrivateTabIndicator` class, then, while the color of the tab indicator in the picker component remains exactly the way we want it (white in dark theme and black in light theme), this behavior also replicates in other places like `/auth/scopes`, thereby fixing the issue I had pointed out in my comment [here](https://github.com/taskcluster/taskcluster/pull/2106#discussion_r360624987).

In other words, whereas all tab indicators **currently** have the same color (`#4177A5`) irrespective of the theme, as shown in [this comment](https://github.com/taskcluster/taskcluster/pull/2106#discussion_r361039726), with this change, the color should be as follows -

`auth/scopes` in Dark theme -
![Screenshot 2019-12-24 21 12 58](https://user-images.githubusercontent.com/11348778/71428985-413fc380-2692-11ea-918f-d8ba280d3049.png)

`auth/scopes` in Light theme -
![Screenshot 2019-12-24 21 12 49](https://user-images.githubusercontent.com/11348778/71429010-5caace80-2692-11ea-9b01-c2adf4867185.png)

I wanted to check with you if this change is acceptable. I happened to come across this behavior while I was actually looking to target just the `PrivateTabIndicator` class in the picker component, using descendant/child class selectors, but that didn't work out (I'll go into more details on that if you want me to).

helfi92 replied:
What do you think of the following change?

```jsx
      MuiPickerDTTabs: {
        tabs: {
          // ...
          '& .MuiTabs-indicator': {
            backgroundColor: isDarkTheme ? '#fff' : '#000',
          },
        },
      },
```

The diff would be:

```diff
diff --git a/ui/src/theme.js b/ui/src/theme.js
index a7f054b8a..2ef178f92 100644
--- a/ui/src/theme.js
+++ b/ui/src/theme.js
@@ -360,14 +360,9 @@ const createTheme = isDarkTheme => {
         tabs: {
           color: '#fff',
           backgroundColor: THEME.SECONDARY,
-        },
-      },
-      PrivateTabIndicator: {
-        colorPrimary: {
-          backgroundColor: '#fff',
-        },
-        colorSecondary: {
-          backgroundColor: '#000',
+          '& .MuiTabs-indicator': {
+            backgroundColor: isDarkTheme ? '#fff' : '#000',
+          },
         },
       },
       MuiListItem: {

```

Manish-Giri replied:
Thank You for the confirmation. I also had something similar, only I didn't think of using `.MuiPickerDTTabs-tabs` as the parent selector!

![Screenshot 2019-12-27 18 33 47](https://user-images.githubusercontent.com/11348778/71536118-f31d0100-28d9-11ea-98e6-9fbb59a2f51c.png)

I'll make the change.

helfi92 commented on ui/src/theme.js:
```diff
@@ -335,6 +335,42 @@ const createTheme = isDarkTheme => {
+        },
+        thumb: {
+          borderColor: THEME.SECONDARY,
+          backgroundColor: 'rgba(254,255, 255, 0.9)',
```
Was there a reason to hard-code this value?

Manish-Giri replied:
Um, the `254` in the red-component is actually a typo. It should be `255`. I'll address this in my next commit. 

As for the opacity values (`0.9` and `0.54`), I found these values from the material-ui pickers.dev component demos. They use `0.9` for picker text which is currently selected, while everything else is dimmed out using `0.54`.

helfi92 replied:
Ah, ok. We can use `PRIMARY_TEXT_DARK`:

https://github.com/taskcluster/taskcluster/blob/0afbbc46a1cba2d6905c76b5f82392380255f0df/ui/src/utils/constants.js#L48

Manish-Giri replied:
Got it. I'll make this change.

Manish-Giri commented:
@helfi92 I updated the PR. Along with the last few changes you had asked for, I also refactored out the hardcoded value for dimming-out unselected text into a new constant in `utils/constants.js`, so if someone were to replicate the behavior in another component, they wouldn't need to hardcode the value again -

![Screenshot 2019-12-27 19 01 34](https://user-images.githubusercontent.com/11348778/71536308-915d9680-28db-11ea-8b67-381c7ae85326.png)

(I hope that's okay!)

helfi92 commented on ui/src/theme.js:
```diff
@@ -297,7 +297,7 @@ const createTheme = isDarkTheme => {
       MuiPickersToolbarText: {
         toolbarTxt: {
-          color: THEME.PRIMARY_TEXT_DARK,
+          color: THEME.PRIMARY_TEXT_DIMMED,
```
We won't be able to use `THEME.PRIMARY_TEXT_DIMMED` on the light theme outside the toolbar. This is because the toolbar has a fixed background color. I think it's okay to hard-code the value here. We're also doing this in the theme file so it shouldn't be too bad.

Manish-Giri replied:
Got it. I updated the PR.

helfi92 commented on changelog/issue-2031.md:
```diff
@@ -0,0 +1,4 @@
+level: minor
+reference: issue 2031
+---
+Taskcluster UI pickers now feature a time component, so for instance, users can now pick a particular time (hour:minute) on a particular day to set as the expiration time.
```
Taskcluster UI revamped the date picker component to allow selecting the hour and the minute in addition to the date.

Manish-Giri replied:
Done

Manish-Giri commented:
@helfi92 Thank You for being so patent, and guiding me along every step of the way! I learnt a lot 💯

helfi92 commented:
My pleasure! Do you want to work on another issue? :-) Happy to help.

Manish-Giri commented:
I'm working on the deep search fix for #2123 and #2124 currently, but if you are aware of another issue that I can contribute to, I'll be happy to take it on too :)

helfi92 commented:
Oh, that's true. We can worry about next steps after these issues are merged. Looking forward to seeing what you come up with!
